server/rest/src/main/java/org/infinispan/rest/CacheInputStream.java

```diff
@@ -0,0 +1,81 @@
+package org.infinispan.rest;
+
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+import org.infinispan.CacheStream;
+
+/**
+ * An {@link InputStream} that reads from a {@link CacheStream} of byte[] and produces a JSON output.
+ *
+ * @since 10.0
+ */
+public class CacheInputStream extends InputStream {
+   private enum State {BEGIN, START_ITEM, ITEM, END_ITEM, SEPARATOR, END, EOF}
+
+   private static final char STREAM_OPEN_CHAR = '[';
+   private static final char DATA_ENCLOSING_CHAR = '\"';
+   private static final char SEPARATOR = ',';
+   private static final char STREAM_CLOSE_CHAR = ']';
+
+   private final Iterator<?> iterator;
+   private final Stream<?> stream;
+   private final int batchSize;
+
+   private byte[] currentEntry;
+   private int cursor = 0;
+
+   private State state = State.BEGIN;
+
+   public CacheInputStream(CacheStream<?> stream, int batchSize, boolean parallel) {
+      CacheStream<?> cacheStream = parallel ? stream.parallel() : stream;
+      this.batchSize = batchSize;
+      this.stream = cacheStream.distributedBatchSize(batchSize);
+      this.iterator = stream.iterator();
+   }
+
+   @Override
+   public int available() {
+      return currentEntry == null ? 0 : currentEntry.length - cursor * batchSize;
+   }
+
+   @Override
+   public synchronized int read() {
+      boolean hasNext = iterator.hasNext();
```

Comment:
Technically this is invoked for every byte of an entry, not sure if we want to optimize that or not?

Reply:
Also technically this method can block, I am not sure if that is an issue or not.

Reply:
@wburns Overall perf looks good, at least compared to the REST V1 "keys" operation, that blew up the server memory. With this PR, I can retrieve millions of keys in a ~70% filled heap in a few seconds

> Technically this is invoked for every byte of an entry

Yes, but the backing stream is batched, and most of time, each byte is already in memory. When a single batch is consumed, it can block, but then from my understanding Netty will use the info from ```available()``` to understand if there is enough 'non-blocking' reads

https://github.com/netty/netty/blob/fd5316ed6fcde05274df38c26b9481771c182c5c/handler/src/main/java/io/netty/handler/stream/ChunkedStream.java

Reply:
btw, I tested locally only :)

Reply:
>@wburns Overall perf looks good, at least compared to the REST V1 "keys" operation, that blew up the server memory. With this PR, I can retrieve millions of keys in a ~70% filled heap in a few seconds

oh, I wasn't trying to compare to the old implementation :) I know this one would be vastly better. Just an observation.

> Yes, but the backing stream is batched, and most of time, each byte is already in memory. When a single batch is consumed, it can block, but then from my understanding Netty will use the info from available() to understand if there is enough 'non-blocking' reads
>
> https://github.com/netty/netty/blob/fd5316ed6fcde05274df38c26b9481771c182c5c/handler/src/main/java/io/netty/handler/stream/ChunkedStream.java

There is no way for netty to know that our method was blocking though. Most of the time it will be in memory, but it is possible for the `hasNext` to block if we are consuming those entries faster than they are being retrieved (internally the iteration tries to keep ahead by requesting extras).

Reply:
right, so the estimation done on ```available()``` is not that accurate given that in the second batch, for example, it can block due to the read ahead. Any suggestions on how to improve it?

Reply:
I was more asking, because I don't know if we need to delegate such an operation to another thread. The only way to be 100% not blocking requires the new non blocking publisher that I have yet to write :(

Reply:
Maybe it's better to wrap this around another thread for now...

Reply:
The other option is we could leave it as is. This method will be unlikely to block. The only way would be if we have exhausted all of the local entries and remote is being quite slow. Iteration prioritizes entries from remote nodes over local ones, to try to reduce latency between entries (since local ones are non blocking). However if persistence is configured for this cache, it will likely block - since the store will most likely be too slow.

Reply:
> Maybe it's better to wrap this around another thread for now...

...if persistence is configured :)

Reply:
Sounds fine to me.

Reply:
Ok, I created https://issues.jboss.org/browse/ISPN-10661 to address the blocking of the IO threads, as it's a bit more involved